File: core/src/com/qtech/bubbleblaster/screen/EnvLoadScreen.java
package com.qtech.bubbleblaster.screen;

import com.qtech.bubbleblaster.BubbleBlaster;
import com.qtech.bubbleblaster.LoadedGame;
import com.qtech.bubbleblaster.common.InfoTransporter;
import com.qtech.bubbleblaster.common.crash.CrashReport;
import com.qtech.bubbleblaster.common.gametype.AbstractGameType;
import com.qtech.bubbleblaster.common.screen.Screen;
import com.qtech.bubbleblaster.core.common.SavedGame;
import com.qtech.bubbleblaster.core.utils.categories.GraphicsUtils;
import com.qtech.bubbleblaster.environment.Environment;
import com.qtech.bubbleblaster.settings.GameSettings;
import org.apache.commons.io.FileUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Supplier;

/**
 * Environment loading scene.
 * When showing this scene, a new thread will be created and in that thread the loading will be done.
 * The thread is located in the method {@link #init()}.
 * <p>
 * Todo: update docstring.
 *
 * @author Quinten Jungblut
 * @since 1.0.615
 */
public class EnvLoadScreen extends Screen {
    private final boolean create;
    private final SavedGame savedGame;
    private final Supplier<AbstractGameType> gameType;
    private final AtomicReference<LoadedGame> loadedGameReference;
    private final InfoTransporter infoTransporter = new InfoTransporter(this::setDescription);
    private String description = "";

    public EnvLoadScreen(SavedGame savedGame, AtomicReference<LoadedGame> loadedGameReference) {
        this(false, savedGame, null, loadedGameReference);
    }

    public EnvLoadScreen(SavedGame savedGame, @NotNull Supplier<AbstractGameType> gameType, AtomicReference<LoadedGame> loadedGameReference) {
        this(true, savedGame, gameType, loadedGameReference);
    }

    private EnvLoadScreen(boolean create, SavedGame savedGame, @Nullable Supplier<AbstractGameType> gameType, AtomicReference<LoadedGame> loadedGameReference) {
        this.create = create;
        this.savedGame = savedGame;
        this.gameType = gameType;
        this.loadedGameReference = loadedGameReference;
    }

    private void setDescription(String description) {
        this.description = description;
    }

    @Override
    public void init() {
        AbstractGameType gameType;
        try {
            File directory = this.savedGame.getDirectory();
            if (this.create && directory.exists()) {
                FileUtils.deleteDirectory(directory);
            }

            if (!directory.exists()) {
                if (!directory.mkdirs()) {
                    throw new IOException("Creating save folder failed.");
                }
            }

            if (create) {
                this.setDescription("Creating save data...");
                gameType = Objects.requireNonNull(this.gameType).get();
                gameType.dumpDefaultState(this.savedGame, this.infoTransporter);
                gameType.createSaveData(this.savedGame, this.infoTransporter);

                this.setDescription("Loading data...");
            } else {
                this.setDescription("Loading data...");
                gameType = AbstractGameType.loadState(this.savedGame, this.infoTransporter);
            }

            Environment environment = this.game.environment = new Environment(gameType);

            if (create) {
                gameType.init(environment, this.infoTransporter);
                gameType.dumpSaveData(this.savedGame);
            } else {
                gameType.load(environment, this.infoTransporter);
            }

            LoadedGame loadedGame = new LoadedGame(savedGame, game.environment);
            loadedGame.run();

            loadedGameReference.set(loadedGame);
        } catch (Throwable t) {
            CrashReport crashReport = new CrashReport("Save being loaded", t);
            crashReport.add("Save Directory", this.savedGame.getDirectory());
            throw crashReport.getReportedException();
        }

        this.game.displayScene(null);
    }

    /**
     * Renders the environment loading scene.<br>
     * Shows the title in the blue accent color (#00b0ff), and the description in a 50% black color (#7f7f7f).
     *
     * @param game the QBubbles game.
     * @param gg   the graphics 2D processor.
     */
    @Override
    public void render(BubbleBlaster game, GraphicsProcessor gg) {
        gg.setColor(new Color(64, 64, 64));
        gg.fillRect(0, 0, BubbleBlaster.getInstance().getWidth(), BubbleBlaster.getInstance().getHeight());
        if (GameSettings.instance().isTextAntialiasEnabled())
            gg.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);

        gg.setColor(new Color(0, 192, 255));
        GraphicsUtils.drawCenteredString(gg, "Loading Environment...", new Rectangle2D.Double(0, ((double) BubbleBlaster.getInstance().getHeight() / 2) - 24, BubbleBlaster.getInstance().getWidth(), 64d), new Font("Helvetica", Font.PLAIN, 48));
        gg.setColor(new Color(127, 127, 127));
        GraphicsUtils.drawCenteredString(gg, this.description, new Rectangle2D.Double(0, ((double) BubbleBlaster.getInstance().getHeight() / 2) + 40, BubbleBlaster.getInstance().getWidth(), 50d), new Font("Helvetica", Font.PLAIN, 20));
        if (GameSettings.instance().isTextAntialiasEnabled())
            gg.setRenderingHint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON);
    }
}

File: src/main/kotlin/qtech/bubbles/screen/EnvLoadScreen.kt
@file:Suppress("unused")

package qtech.bubbles.screen

import qtech.bubbles.BubbleBlaster
import qtech.bubbles.LoadedGame
import qtech.bubbles.common.InfoTransporter
import qtech.hydro.crash.CrashReport
import qtech.bubbles.common.gametype.AbstractGameMode
import qtech.bubbles.common.gametype.AbstractGameMode.Companion.loadState
import qtech.bubbles.common.screen.Screen
import qtech.bubbles.core.common.SavedGame
import qtech.bubbles.core.utils.categories.GraphicsUtils.drawCenteredString
import qtech.bubbles.environment.Environment
import qtech.hydro.GraphicsProcessor
import qtech.bubbles.settings.GameSettings
import org.apache.commons.io.FileUtils
import qtech.hydro.Game
import qtech.hydro.crash.CrashCategory
import java.awt.Color
import java.awt.Font
import java.awt.RenderingHints
import java.awt.geom.Rectangle2D
import java.io.File
import java.io.IOException
import java.util.*
import java.util.concurrent.atomic.AtomicReference
import java.util.function.Supplier

/**
 * Environment loading scene.
 * When showing this scene, a new thread will be created and in that thread the loading will be done.
 * The thread is located in the method [.init].
 *
 *
 * Todo: update docstring.
 *
 * @author Quinten Jungblut
 * @since 1.0.615
 */
class EnvLoadScreen private constructor(private val create: Boolean, private val savedGame: SavedGame, gameMode: Supplier<AbstractGameMode>?, loadedGameReference: AtomicReference<LoadedGame?>) : Screen() {
    private val gameMode: Supplier<AbstractGameMode>?
    private val loadedGameReference: AtomicReference<LoadedGame?>
    private val infoTransporter = InfoTransporter { description: String -> setDescription(description) }
    private var description = ""

    constructor(savedGame: SavedGame, loadedGameReference: AtomicReference<LoadedGame?>) : this(false, savedGame, null, loadedGameReference)
    constructor(savedGame: SavedGame, gameMode: Supplier<AbstractGameMode>, loadedGameReference: AtomicReference<LoadedGame?>) : this(true, savedGame, gameMode, loadedGameReference)

    private fun setDescription(description: String) {
        this.description = description
    }

    override fun init() {
        val gameMode: AbstractGameMode
        try {
            val directory: File = savedGame.directory
            if (create && directory.exists()) {
                FileUtils.deleteDirectory(directory)
            }
            if (!directory.exists()) {
                if (!directory.mkdirs()) {
                    throw IOException("Creating save folder failed.")
                }
            }
            if (create) {
                setDescription("Creating save data...")
                gameMode = Objects.requireNonNull(this.gameMode)!!.get()
                gameMode.dumpDefaultState(savedGame, infoTransporter)
                gameMode.createSaveData(savedGame, infoTransporter)
                setDescription("Loading data...")
            } else {
                setDescription("Loading data...")
                gameMode = loadState(savedGame, infoTransporter)
            }
            game.environment = Environment(gameMode)
            val environment = game.environment!!
            if (create) {
                gameMode.init(environment, infoTransporter)
                gameMode.dumpSaveData(savedGame)
            } else {
                gameMode.load(environment, infoTransporter)
            }
            val loadedGame = LoadedGame(savedGame, game.environment!!)
            loadedGame.run()
            loadedGameReference.set(loadedGame)
        } catch (t: Throwable) {
            val crashReport = CrashReport(t)
            crashReport.addCategory(CrashCategory("Environment being loaded").also {
                it.add("Save Directory") { savedGame.directory }
            })
            throw crashReport.reportedException
        }
        game.displayScene(null)
    }

    /**
     * Renders the environment loading scene.<br></br>
     * Shows the title in the blue accent color (#00b0ff), and the description in a 50% black color (#7f7f7f).
     * @param game the QBubbles game.
     * @param gp   the graphics 2D processor.
     */
    override fun render(game: Game, gp: GraphicsProcessor) {
        gp.color(Color(64, 64, 64))
        gp.rectF(0, 0, BubbleBlaster.instance.width, BubbleBlaster.instance.height)
        if (GameSettings.instance().isTextAntialiasEnabled) gp.hint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON)
        gp.color(Color(0, 192, 255))
        drawCenteredString(gp, "Loading Environment...", Rectangle2D.Double(0.0, BubbleBlaster.instance.height.toDouble() / 2 - 24, BubbleBlaster.instance.width.toDouble(), 64.0), Font("Helvetica", Font.PLAIN, 48))
        gp.color(Color(127, 127, 127))
        drawCenteredString(gp, description, Rectangle2D.Double(0.0, BubbleBlaster.instance.height.toDouble() / 2 + 40, BubbleBlaster.instance.width.toDouble(), 50.0), Font("Helvetica", Font.PLAIN, 20))
        if (GameSettings.instance().isTextAntialiasEnabled) gp.hint(RenderingHints.KEY_TEXT_ANTIALIASING, RenderingHints.VALUE_TEXT_ANTIALIAS_ON)
    }

    init {
        this.gameMode = gameMode
        this.loadedGameReference = loadedGameReference
    }
}